src/secret_connection.rs

```diff
@@ -0,0 +1,457 @@
+
+impl<IoHandler: io::Read + io::Write + Send + Sync> SecretConnection<IoHandler> {
+    // Returns authenticated remote pubkey
+    fn remote_pubkey(&self) -> [u8; 32] {
+        self.remote_pubkey
+    }
+    // Performs handshake and returns a new authenticated SecretConnection.
+    pub fn new(
+        mut handler: IoHandler,
+        local_privkey: &DalekSigner,
+    ) -> Result<SecretConnection<IoHandler>, Error> {
+        // TODO: Error check
+        let local_pubkey = local_privkey.public_key().unwrap();
+
+        // Generate ephemeral keys for perfect forward secrecy.
+        let (local_eph_pubkey, local_eph_privkey) = gen_eph_keys();
+
+        // Write local ephemeral pubkey and receive one too.
+        // NOTE: every 32-byte string is accepted as a Curve25519 public key
+        // (see DJB's Curve25519 paper: http://cr.yp.to/ecdh/curve25519-20060209.pdf)
+        let remote_eph_pubkey = share_eph_pubkey(&mut handler, &local_eph_pubkey).unwrap();
+
+        // Compute common shared secret.
+        let shared_secret = diffie_hellman(&local_eph_privkey, &remote_eph_pubkey);
+
+        // Sort by lexical order.
+        let (low_eph_pubkey, _) = sort32(local_eph_pubkey, remote_eph_pubkey);
+
+        // Check if the local ephemeral public key
+        // was the least, lexicographically sorted.
+        let loc_is_least = local_eph_pubkey == low_eph_pubkey;
+
+        let (recv_secret, send_secret, challenge) =
+            derive_secrets_and_challenge(&shared_secret, loc_is_least);
+
+        // Construct SecretConnection.
+        let mut sc = SecretConnection {
+            io_handler: handler,
+            recv_buffer: vec![],
+            recv_nonce: [0u8; AEAD_NONCE_SIZE],
+            send_nonce: [0u8; AEAD_NONCE_SIZE],
+            recv_secret: aead::OpeningKey::new(&aead::CHACHA20_POLY1305, &recv_secret).unwrap(),
+            send_secret: aead::SealingKey::new(&aead::CHACHA20_POLY1305, &send_secret).unwrap(),
+            remote_pubkey: remote_eph_pubkey,
+        };
+
+        // Sign the challenge bytes for authentication.
+        // TODO: Error check
+        let local_signature = sign_challenge(challenge, local_privkey).unwrap();
+
+        // Share (in secret) each other's pubkey & challenge signature
+        // TODO: Error check
+        let auth_sig_msg =
+            share_auth_signature(&mut sc, local_pubkey.as_bytes(), local_signature).unwrap();
+
+        let remote_pubkey = PublicKey::from_bytes(&auth_sig_msg.key).unwrap();
+        let remote_signature: &[u8] = &auth_sig_msg.sig;
+        let remote_sig = Signature::from_bytes(remote_signature).unwrap();
+
+        let valid_sig = DefaultVerifier::verify(&remote_pubkey, &challenge, &remote_sig);
+
+        valid_sig.map_err(|e| err!(ChallengeVerification, "{}", e))?;
+
+        // We've authorized.
+        sc.remote_pubkey.copy_from_slice(&auth_sig_msg.key);
+        Ok(sc)
+    }
+}
+
+fn open(
+    opening_key: &aead::OpeningKey,
+    nonce: &[u8; 12],
```

Comment:
Given `opening_key` and `nonce` both come from `SecretConnection`, any reason to not make this a private method on `SecretConnection`? It looks like a corresponding `seal()` method would be useful too.